### PR TITLE
[Crypto] general updates

### PR DESCRIPTION
**Pull request by tarakby**
## Description

- make changes to the kms signer `Sign` function so that hashing the message is done by the cloud KMS.
- add Keccak256 to the package supported algorithms.
- tidy up the ECDSA curves/hash functions compatibility. In particular, separate compatibility for Flow account keys and compatibility of using ECDSA.
- Closes: #193 (tidy up the support of secp256k1 by the kms signer).
- add a new `PublicKey` method to the `Signer` interface. 

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-go-sdk/blob/master/CONTRIBUTING.md#styleguides).
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: crypto/hash.go
Comment (tarakby) [inline crypto/hash.go:102]: I was wondering if I can update all the functions of type :
```
func SdkFunction() {
    flow_go_crypto.Funtion()
}
```
to 
```
var  SdkFunction = flow_go_crypto.Funtion
```
Is there a reason the first type was preferred so far?

File: crypto/crypto.go
Comment (tarakby) [inline crypto/crypto.go:56]: After moving the hashing related stuff to `hash.go`, this function only contains ECDSA signing functions. I would like to rename the file to `sign.go`. I will proceed to rename it later to not complicate the review. Let me know if there is a concern with the renaming.

File: crypto/cloudkms/signer.go
Comment (tarakby) [inline crypto/cloudkms/signer.go:76]: This change was not tested and requires testing with a KMS. How do you usually test these KMS changes ?  @psiemens @judezhu 
Ideally, the test would be added to CI (maybe have a kms key (ring) accessible by CI if that's simple) ?

File: account.go
Comment (sideninja) [inline account.go:107]: nit: isn't this a bit confusing? although it's technically correct in this case it just replaces the `return true` in the next case and I wonder if that makes it better? I feel like it would be more clear if it would just return true.

File: account_test.go
Comment (sideninja) [inline account_test.go:108]: nice 👍

File: crypto/cloudkms/signer.go
Comment (sideninja) [reply crypto/cloudkms/signer.go:76]: I was testing that manually before. You need to have access to KMS (contact SRE) and then create a key (there should also be one already created by me a while ago). Like you said adding this to the CI and making an automatic test for it would be way better 💯

File: crypto/cloudkms/signer.go
Comment (sideninja) [inline crypto/cloudkms/signer.go:130]: panic would probably be better

File: crypto/hash.go
Comment (sideninja) [reply crypto/hash.go:102]: I don't know the initial design decisions (so hopefully Pete will speak to that) but this proposed change I guess has a downside of the possibility to overwrite that value saved there. Also, the call stack would be different but that I don't think has much of purpose anyway.

**Review by sideninja — APPROVED** [review]
Great improvements, thanks. LGTM 🚀

File: crypto/cloudkms/signer.go
Comment (judezhu) [reply crypto/cloudkms/signer.go:76]: hey @tarakby, yeah! same here. In my local testing, we were using a `kms` key hosted on GCP. You probably can mock the `kms` sign function: `AsymmetricSign`, if you wanna get rid of the `kms` dependency in our local testing here.

**Comment by LanfordCai** [comment]
I'm wondering maybe you will be interested in this: https://github.com/onflow/flow-go/issues/2251. It's also related to flow-go-sdk

**Comment by tarakby** [comment]
This PR isn't completed (to address comments) and isn't fully tested yet (kms new signing request). I just didn't get time to get back to it recently but I should work on it next week. Is reverting the merge possible?

**Comment by sideninja** [comment]
Yes, it is. Sorry for merging but if the PR is open and has approval I assume to merge it. Maybe we need to be more clear about who is also merging them in the future.

**Comment by tarakby** [comment]
Agree, I should have said clearly that I'm adding more changes, or blocked merging. 
In `flow-go`, we are used to let the PR creator merge when they get approvals and think it's ready. I assumed the same for `flow-go-sdk`.

**Comment by sideninja** [comment]
Yeah, no worries @tarakby it's also partially my fault for not clearing it with you. There's no protocol for go-sdk since there's only me actively maintaining but what you said makes sense.

File: crypto/hash.go
Comment (tarakby) [reply crypto/hash.go:102]: Overwriting is a good point, though I think there is a low risk for a developer to overwrite a that value. 
For the call stack, since this is a short function, the complier might be already inlining it. 

I'll just keep the original code :)

Comment (sideninja) [reply crypto/hash.go:102]: Honestly, I'm fine either way, I just wanted to provide my answer to the question.

File: account.go
Comment (tarakby) [reply account.go:107]: I totally agree 😄 the `fallthrough` was in the original code and I kept it. I'll update it.

**Pull request merged**
